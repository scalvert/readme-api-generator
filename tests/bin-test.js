import path from 'path';
import fs from 'fs';
import { describe, beforeEach, afterEach, it, expect } from 'vitest';
import { Project } from 'fixturify-project';
import execa from 'execa';

class FakeProject extends Project {
  write(dirJSON) {
    Object.assign(this.files, dirJSON);
    this.writeSync();
  }
}

describe('readme-api-generator', () => {
  let project;

  beforeEach(() => {
    project = new FakeProject();

    project.writeSync();
  });

  afterEach(() => {
    project.dispose();
  });

  it('returns error if no README is present', async () => {
    let result = await run();

    expect(result.stderr).toMatchInlineSnapshot(
      `"No README found. Please run readme-api-generator from a directory that contains a README.md file."`
    );
    expect(result.exitCode).toEqual(1);
  });

  it('returns error if README does not contain start or end placeholders', async () => {
    project.write({
      'README.md': '',
    });

    let result = await run();

    expect(result.stderr).toMatchInlineSnapshot(
      `"The README does not contain a valid placeholder (<!--DOCS_START--> followed by a <!--DOCS_END--> HTML comment)"`
    );
    expect(result.exitCode).toEqual(1);
  });

  it('returns error if README does not contain start placeholder', async () => {
    project.write({
      'README.md': '<!--DOCS_END-->',
    });

    let result = await run();

    expect(result.stderr).toMatchInlineSnapshot(
      `"The README does not contain a valid placeholder (<!--DOCS_START--> followed by a <!--DOCS_END--> HTML comment)"`
    );
    expect(result.exitCode).toEqual(1);
  });

  it('returns error if README does not contain end placeholder', async () => {
    project.write({
      'README.md': '<!--DOCS_START-->',
    });

    let result = await run();

    expect(result.stderr).toMatchInlineSnapshot(
      `"The README does not contain a valid placeholder (<!--DOCS_START--> followed by a <!--DOCS_END--> HTML comment)"`
    );
    expect(result.exitCode).toEqual(1);
  });

  it('writes content for single file', async () => {
    project.write({
      'README.md': `Fake readme
<!--DOCS_START--><!--DOCS_END-->`,
      'protection.js': `/**
 * A quite wonderful function.
 * @param {object} - Privacy gown
 * @param {object} - Security
 * @returns {survival}
 */
function protection (cloak, dagger) {}
`,
    });

    let result = await run(['protection.js']);

    expect(result.stdout).toMatchInlineSnapshot(`"README content updated"`);
    expect(
      fs.readFileSync(path.join(project.baseDir, 'README.md'), {
        encoding: 'utf-8',
      })
    ).toMatchInlineSnapshot();
    expect(result.exitCode).toEqual(0);
  });

  it('writes content for multiple files', async () => {
    project.write({
      'README.md': `Fake readme
<!--DOCS_START--><!--DOCS_END-->`,
      'protection.js': `/**
 * A quite wonderful function.
 * @param {object} - Privacy gown
 * @param {object} - Security
 * @returns {survival}
 */
function protection (cloak, dagger) {}
`,
      'defense.js': `/**
 * Another wonderful function.
 * @param {object} - Privacy gown
 * @param {object} - Security
 * @returns {alive}
 */
function defense (cloak2, dagger2) {}
`,
    });

    let result = await run(['protection.js', 'defense.js']);

    expect(result.stdout).toMatchInlineSnapshot(`
      "Fake readme
      <!--DOCS_START-->
      <a name=\\"protection\\"></a>
      
      ## protection(cloak, dagger) ⇒ <code>survival</code>
      A quite wonderful function.
      
      **Kind**: global function  
      
      | Param | Type | Description |
      | --- | --- | --- |
      | cloak | <code>object</code> | Privacy gown |
      | dagger | <code>object</code> | Security |
      
      
      <!--DOCS_END-->"
    `);
    expect(
      fs.readFileSync(path.join(project.baseDir, 'README.md'), {
        encoding: 'utf-8',
      })
    ).toMatchInlineSnapshot();
    expect(result.exitCode).toEqual(0);
  });

  it('writes content for directories', async () => {
    project.write({
      'README.md': `Fake readme
<!--DOCS_START--><!--DOCS_END-->`,
      lib: {
        'protection.js': `/**
   * A quite wonderful function.
   * @param {object} - Privacy gown
   * @param {object} - Security
   * @returns {survival}
   */
  function protection (cloak, dagger) {}
  `,
        'defense.js': `/**
   * Another wonderful function.
   * @param {object} - Privacy gown
   * @param {object} - Security
   * @returns {alive}
   */
  function defense (cloak2, dagger2) {}
  `,
      },
    });

    let result = await run(['lib']);

    expect(result.stdout).toMatchInlineSnapshot(`
      "Fake readme
      <!--DOCS_START-->
      ## Functions
      
      <dl>
      <dt><a href=\\"#protection\\">protection(cloak, dagger)</a> ⇒ <code>survival</code></dt>
      <dd><p>A quite wonderful function.</p>
      </dd>
      <dt><a href=\\"#defense\\">defense(cloak2, dagger2)</a> ⇒ <code>alive</code></dt>
      <dd><p>Another wonderful function.</p>
      </dd>
      </dl>
      
      <a name=\\"protection\\"></a>
      
      ## protection(cloak, dagger) ⇒ <code>survival</code>
      A quite wonderful function.
      
      **Kind**: global function  
      
      | Param | Type | Description |
      | --- | --- | --- |
      | cloak | <code>object</code> | Privacy gown |
      | dagger | <code>object</code> | Security |
      
      <a name=\\"defense\\"></a>
      
      ## defense(cloak2, dagger2) ⇒ <code>alive</code>
      Another wonderful function.
      
      **Kind**: global function  
      
      | Param | Type | Description |
      | --- | --- | --- |
      | cloak2 | <code>object</code> | Privacy gown |
      | dagger2 | <code>object</code> | Security |
      
      
      <!--DOCS_END-->"
    `);
    expect(
      fs.readFileSync(path.join(project.baseDir, 'README.md'), {
        encoding: 'utf-8',
      })
    ).toMatchInlineSnapshot();
    expect(result.exitCode).toEqual(0);
  });

  it('can use custom markers', async () => {
    project.write({
      'README.md': `Fake readme
<!--CUSTOM_START--><!--CUSTOM_END-->`,
      'protection.js': `/**
 * A quite wonderful function.
 * @param {object} - Privacy gown
 * @param {object} - Security
 * @returns {survival}
 */
function protection (cloak, dagger) {}
`,
    });

    let result = await run(['protection.js', '--marker', 'CUSTOM']);

    expect(result.stdout).toMatchInlineSnapshot(`
      "Fake readme
      <!--DOCS_START-->
      ## Functions
      
      <dl>
      <dt><a href=\\"#defense\\">defense(cloak2, dagger2)</a> ⇒ <code>alive</code></dt>
      <dd><p>Another wonderful function.</p>
      </dd>
      <dt><a href=\\"#protection\\">protection(cloak, dagger)</a> ⇒ <code>survival</code></dt>
      <dd><p>A quite wonderful function.</p>
      </dd>
      </dl>
      
      <a name=\\"defense\\"></a>
      
      ## defense(cloak2, dagger2) ⇒ <code>alive</code>
      Another wonderful function.
      
      **Kind**: global function  
      
      | Param | Type | Description |
      | --- | --- | --- |
      | cloak2 | <code>object</code> | Privacy gown |
      | dagger2 | <code>object</code> | Security |
      
      <a name=\\"protection\\"></a>
      
      ## protection(cloak, dagger) ⇒ <code>survival</code>
      A quite wonderful function.
      
      **Kind**: global function  
      
      | Param | Type | Description |
      | --- | --- | --- |
      | cloak | <code>object</code> | Privacy gown |
      | dagger | <code>object</code> | Security |
      
      
      <!--DOCS_END-->"
    `);
    expect(
      fs.readFileSync(path.join(project.baseDir, 'README.md'), {
        encoding: 'utf-8',
      })
    ).toMatchInlineSnapshot();
    expect(result.exitCode).toEqual(0);
  });

  function run(args = [], options = {}) {
    let defaults = {
      reject: false,
      cwd: project.baseDir,
    };

    return execa(
      process.execPath,
      [require.resolve('../bin/readme-api-generator.js'), ...args],
      Object.assign({}, defaults, options)
    );
  }
});
