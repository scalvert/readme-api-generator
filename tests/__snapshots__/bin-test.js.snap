// Vitest Snapshot v1

exports[`readme-api-generator > can use custom markers 2`] = `
"Fake readme
<!--CUSTOM_START-->
<a name=\\"protection\\"></a>

## protection(cloak, dagger) ⇒ <code>survival</code>
A quite wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak | <code>object</code> | Privacy gown |
| dagger | <code>object</code> | Security |


<!--CUSTOM_END-->"
`;

exports[`readme-api-generator > writes content for directories 2`] = `
"Fake readme
<!--DOCS_START-->
## Functions

<dl>
<dt><a href=\\"#defense\\">defense(cloak2, dagger2)</a> ⇒ <code>alive</code></dt>
<dd><p>Another wonderful function.</p>
</dd>
<dt><a href=\\"#protection\\">protection(cloak, dagger)</a> ⇒ <code>survival</code></dt>
<dd><p>A quite wonderful function.</p>
</dd>
</dl>

<a name=\\"defense\\"></a>

## defense(cloak2, dagger2) ⇒ <code>alive</code>
Another wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak2 | <code>object</code> | Privacy gown |
| dagger2 | <code>object</code> | Security |

<a name=\\"protection\\"></a>

## protection(cloak, dagger) ⇒ <code>survival</code>
A quite wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak | <code>object</code> | Privacy gown |
| dagger | <code>object</code> | Security |


<!--DOCS_END-->"
`;

exports[`readme-api-generator > writes content for multiple files 2`] = `
"Fake readme
<!--DOCS_START-->
## Functions

<dl>
<dt><a href=\\"#protection\\">protection(cloak, dagger)</a> ⇒ <code>survival</code></dt>
<dd><p>A quite wonderful function.</p>
</dd>
<dt><a href=\\"#defense\\">defense(cloak2, dagger2)</a> ⇒ <code>alive</code></dt>
<dd><p>Another wonderful function.</p>
</dd>
</dl>

<a name=\\"protection\\"></a>

## protection(cloak, dagger) ⇒ <code>survival</code>
A quite wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak | <code>object</code> | Privacy gown |
| dagger | <code>object</code> | Security |

<a name=\\"defense\\"></a>

## defense(cloak2, dagger2) ⇒ <code>alive</code>
Another wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak2 | <code>object</code> | Privacy gown |
| dagger2 | <code>object</code> | Security |


<!--DOCS_END-->"
`;

exports[`readme-api-generator > writes content for single file 2`] = `
"Fake readme
<!--DOCS_START-->
<a name=\\"protection\\"></a>

## protection(cloak, dagger) ⇒ <code>survival</code>
A quite wonderful function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| cloak | <code>object</code> | Privacy gown |
| dagger | <code>object</code> | Security |


<!--DOCS_END-->"
`;
